In [7]:
import numpy as np
import matplotlib.pyplot as plt

ModuleNotFoundError: No module named 'matplotlib'

# Regression Coefficients
Construct linear regression using 3 methods:
1. Least Square
2. Gradient Descent
3. Stochastic Gradient Descent: SGD

Example: Given Heights and Weights Dataset of 9 people as follows:
X = ([152, 157, 160, 163, 150, 147, 165, 168, 170])
y = ([54.48, 55.84, 57.20, 56.57, 53.12, 52.21, 59.93, 61.29, 67.92])


In [ ]:
X = np.array([152, 157, 160, 163, 150, 147, 165, 168, 170])
Y = np.array([54.48, 55.84, 57.20, 56.57, 53.12, 52.21, 59.93, 61.29, 67.92])

## 1. Least Square
To construct the regression coefficients using the Least Squares Method, we follow these steps:

1. **Define the linear regression model**: $ y = \beta_0 + \beta_1 x $

2. **Calculate the means of $X$ and $y$**:
   $$
   \bar{X} = \frac{1}{n} \sum_{i=1}^n X_i \quad \text{and} \quad \bar{y} = \frac{1}{n} \sum_{i=1}^n y_i
   $$
   where $n$ is the number of data points.

3. **Calculate the regression coefficients**:
   $$
   \beta_1 = \frac{\sum_{i=1}^n (X_i - \bar{X})(y_i - \bar{y})}{\sum_{i=1}^n (X_i - \bar{X})^2}
   $$
   $$
   \beta_0 = \bar{y} - \beta_1 \bar{X}
   $$

Given data:
$ X = [152, 157, 160, 163, 150, 147, 165, 168, 170] $
$ y = [54.48, 55.84, 57.20, 56.57, 53.12, 52.21, 59.93, 61.29, 67.92] $

### Step 1: Calculate the means of $X$ and $y$
$$
\bar{X} = \frac{152 + 157 + 160 + 163 + 150 + 147 + 165 + 168 + 170}{9}
$$


In [6]:
barX = np.sum(X) / 9

NameError: name 'np' is not defined

$$
\bar{y} = \frac{54.48 + 55.84 + 57.20 + 56.57 + 53.12 + 52.21 + 59.93 + 61.29 + 67.92}{9}
$$

In [5]:
(54.48 + 55.84 + 57.20 + 56.57 + 53.12 + 52.21 + 59.93 + 61.29 + 67.92) / 9

57.617777777777775

### Step 2: Calculate the regression coefficients
First, we calculate the numerator and denominator for $\beta_1$:
$$
\sum_{i=1}^n (X_i - \bar{X})(y_i - \bar{y}) = (152-160)(54.48-57.618) + (157-160)(55.84-57.618) + (160-160)(57.20-57.618) + \cdots + (170-160)(67.92-57.618)
$$

### Step 2: Calculate the regression coefficients
First, we calculate the numerator and denominator for $\beta_1$:
$$
\sum_{i=1}^n (X_i - \bar{X})(y_i - \bar{y}) = (152-160)(54.48-57.618) + (157-160)(55.84-57.618) + (160-160)(57.20-57.618) + \cdots + (170-160)(67.92-57.618)
$$

Calculate each term:
$$
(152 - 160)(54.48 - 57.618) = (-8)(-3.138) = 25.104
$$
$$
(157 - 160)(55.84 - 57.618) = (-3)(-1.778) = 5.334
$$
$$
(160 - 160)(57.20 - 57.618) = (0)(-0.418) = 0
$$
$$
(163 - 160)(56.57 - 57.618) = (3)(-1.048) = -3.144
$$
$$
(150 - 160)(53.12 - 57.618) = (-10)(-4.498) = 44.98
$$
$$
(147 - 160)(52.21 - 57.618) = (-13)(-5.408) = 70.304
$$
$$
(165 - 160)(59.93 - 57.618) = (5)(2.312) = 11.56
$$
$$
(168 - 160)(61.29 - 57.618) = (8)(3.672) = 29.376
$$
$$
(170 - 160)(67.92 - 57.618) = (10)(10.302) = 103.02
$$

Sum of products:
$$
25.104 + 5.334 + 0 - 3.144 + 44.98 + 70.304 + 11.56 + 29.376 + 103.02 = 286.534
$$

Now, calculate the denominator:
$$
\sum_{i=1}^n (X_i - \bar{X})^2 = (152-160)^2 + (157-160)^2 + (160-160)^2 + \cdots + (170-160)^2
$$
$$
(152 - 160)^2 = (-8)^2 = 64
$$
$$
(157 - 160)^2 = (-3)^2 = 9
$$
$$
(160 - 160)^2 = 0
$$
$$
(163 - 160)^2 = 3^2 = 9
$$
$$
(150 - 160)^2 = (-10)^2 = 100
$$
$$
(147 - 160)^2 = (-13)^2 = 169
$$
$$
(165 - 160)^2 = 5^2 = 25
$$
$$
(168 - 160)^2 = 8^2 = 64
$$
$$
(170 - 160)^2 = 10^2 = 100
$$

Sum of squares:
$$
64 + 9 + 0 + 9 + 100 + 169 + 25 + 64 + 100 = 540
$$

Calculate $\beta_1$:
$$
\beta_1 = \frac{286.534}{540} \approx 0.5306
$$

Calculate $\beta_0$:
$$
\beta_0 = \bar{y} - \beta_1 \bar{X} = 57.618 - (0.5306 \times 160) = 57.618 - 84.896 = -27.278
$$

### Step 3: Regression equation
The regression equation is:
$$
y = -27.278 + 0.5306x
$$

Therefore, the regression coefficients are $\beta_0 \approx -27.278$ and $\beta_1 \approx 0.5306$.